cmd/cleanupexamples/main.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2024 Upbound Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// main package for the cleanupexamples tooling, the tool to remove
+// uptest-specific code from published examples on the marketplace
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	kyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/yaml"
+)
+
+// removeAnnotations removes specific annotations from a Kubernetes
+// unstructured object. It looks for annotations with prefixes
+// "upjet.upbound.io/" and "uptest.upbound.io/" and removes
+// them if they are present.
+func removeAnnotations(u *unstructured.Unstructured) {
+	annotations := u.GetAnnotations()
+	if annotations != nil {
+		annotationsToRemove := []string{
+			"upjet.upbound.io/",
+			"uptest.upbound.io/",
+		}
+
+		for key := range annotations {
+			for _, prefix := range annotationsToRemove {
+				if strings.HasPrefix(key, prefix) {
+					delete(annotations, key)
+					break
+				}
+			}
+		}
+
+		if len(annotations) == 0 {
+			u.SetAnnotations(nil)
+		} else {
+			u.SetAnnotations(annotations)
+		}
+	}
+}
+
+// processYAML processes a YAML file by replacing specific placeholders
+// and removing certain annotations from the Kubernetes objects within it.
+// It returns the modified YAML content as a byte slice with YAML document
+// separators ("---") between each object.
+func processYAML(yamlData []byte) ([]byte, error) {
+	// TODO(turkenf): Handle replacing UPTEST_DATASOURCE placeholders like ${data.aws_account_id}
+	yamlData = bytes.ReplaceAll(yamlData, []byte("${Rand.RFC1123Subdomain}"), []byte("random"))
+
+	// Create a YAML or JSON decoder to read and decode the input YAML data
+	decoder := kyaml.NewYAMLOrJSONDecoder(bytes.NewReader(yamlData), 1024)
+	var modifiedYAMLs []byte
+	separator := []byte("---\n")
+	first := true
+
+	for {
+		u := &unstructured.Unstructured{}
+		if err := decoder.Decode(&u); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, fmt.Errorf("cannot decode the YAML file: %w", err)
+		}
+
+		// Remove specific annotations from the decoded Kubernetes object.
+		removeAnnotations(u)
```

Comment:
Thought: Does `filterAnnotations()` sound better than `removeAnnotations()`? The latter sounds to me like it will remove all annotations.

Reply:
Done.